pkg/agent/cache/cache.go

```diff
@@ -42,18 +44,43 @@ func NewCache() *cacheImpl {
 	return &cacheImpl{cache: make(map[string][]CacheEntry)}
 }
 
+func (c *cacheImpl) Entries() map[string][]CacheEntry {
+	c.m.Lock()
+	defer c.m.Unlock()
+	return c.cache
+
+}
+
 func (c *cacheImpl) Entry(selectors []*common.Selector) (entry []CacheEntry) {
 	key := deriveCacheKey(selectors)
 	c.m.Lock()
 	defer c.m.Unlock()
-	return c.cache[key]
+	if entry, found := c.cache[key]; found {
+		return entry
+	}
+	return nil
 }
 
 func (c *cacheImpl) SetEntry(cacheEntry CacheEntry) {
 	c.m.Lock()
 	defer c.m.Unlock()
 	key := deriveCacheKey(cacheEntry.RegistrationEntry.Selectors)
+	fmt.Println(len(c.cache))
+	fmt.Println(len(c.cache[key]))
+
+	for i, entry := range c.cache[key] {
+		fmt.Println(entry)
+		fmt.Println(cacheEntry)
+		if entry.RegistrationEntry.SpiffeId == cacheEntry.RegistrationEntry.SpiffeId {
+			copy(c.cache[key][i:], c.cache[key][i+1:])
+			c.cache[key][len(c.cache[key])-1] = CacheEntry{}
+			c.cache[key] = c.cache[key][:len(c.cache[key])-1]
```

Comment:
non-blocking: this is pretty hard to read. maybe a comment or something will help